File: src/Solutions/Euler193.hs
module Solutions.Euler193 where
import Utils.NumberTheory(unPrime, primesUpTo)

-- Inclusion–exclusion principle

compute :: Integer -> Integer
compute bound2 =
    let bound = floor $ sqrt $ fromIntegral bound2
        getNumbers :: Integer -> [Integer] -> Integer
        getNumbers prod [] = bound2 `quot` prod
        getNumbers prod (p:ps)
            | abs (p * prod) >= bound2 = bound2 `quot` prod
            | p >= bound2 = 0
            | otherwise = getNumbers prod ps + getNumbers ((-p) * prod) ps
    in getNumbers 1 $ map ((^2) . unPrime) $ primesUpTo bound

euler193 :: IO String
euler193 = return $ show $ compute $ 2^50

File: rust/src/solutions/euler193.rs
// Inclusion-exclusion principle

fn count_squarefrees(bound: i64) -> i64 {
    primal::Primes::all()
        .map(|p| (p * p) as i64)
        .take_while(|&p| p < bound)
        .fold((bound - 1, vec![1i64]), |(mut sum, prods), p| {
            let mut next = Vec::new();
            for prod in prods {
                if prod
                    .checked_mul(p)
                    .map(|x| x.abs() >= bound)
                    .unwrap_or(true)
                {
                    sum += bound / prod;
                } else {
                    next.push(prod);
                    next.push(-p * prod);
                }
            }
            (sum, next)
        })
        .0
}

pub fn euler193() -> String {
    count_squarefrees(2i64.pow(50)).to_string()
}
